File: trees/543_diameter_of_binary_tree.py
from typing import Optional


"""
Given the root of a binary tree, return the length of the diameter of the tree.

The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

The length of a path between two nodes is represented by the number of edges between them.

https://leetcode.com/problems/diameter-of-binary-tree/description/
O(n), O(n)
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right=right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        dia = 0

        def dfs(root):
            nonlocal dia
            if not root:
                return 0
            left = dfs(root.left)
            right = dfs(root.right)
            dia = max(dia, left + right)
            return 1 + max(left, right)

        dfs(root)
        return dia


if __name__ == "__main__":
    root = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3))
    print(Solution().diameterOfBinaryTree(root))

File: trees/543_diameter_of_binary_tree.cpp
#include <iostream>
using namespace std;

/* Given the root of a binary tree, return the length of the diameter of the tree.

The diameter of a binary tree is the length of the longest path between any two nodes in a tree. This path may or may not pass through the root.

The length of a path between two nodes is represented by the number of edges between them.

https://leetcode.com/problems/diameter-of-binary-tree/description/
O(n), O(n) */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};

class Solution {
public:
    int diameterOfBinaryTree(TreeNode *root) {
        int dia = 0;
        dfs(root, dia);
        return dia;
    }

private:
    int dfs(TreeNode* root, int& dia) {
        if (root == nullptr) {
            return 0;
        }
        int left = dfs(root->left, dia);
        int right = dfs(root->right, dia);
        dia = max(dia, left+right);
        return 1 + max(left, right); // height
    }
};

int main(int argc, char *argv[]) {
    TreeNode *root = new TreeNode(
        1, new TreeNode(2, new TreeNode(4), new TreeNode(5)), new TreeNode(3));
    cout << Solution().diameterOfBinaryTree(root) << endl;
    return 0;
}
